java/arcs/android/util/ProtoPrefetcher.kt

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+package arcs.android.util
+
+import android.os.Trace
+import arcs.android.crdt.CrdtCountProto
+import arcs.android.crdt.CrdtDataProto
+import arcs.android.crdt.CrdtEntityProto
+import arcs.android.crdt.CrdtEntityReferenceProto
+import arcs.android.crdt.CrdtOperationProto
+import arcs.android.crdt.CrdtSetProto
+import arcs.android.crdt.CrdtSingletonProto
+import arcs.android.crdt.RawEntityProto
+import arcs.android.crdt.ReferencablePrimitiveProto
+import arcs.android.crdt.ReferencableProto
+import arcs.android.crdt.ReferencableSetProto
+import arcs.android.crdt.ReferenceProto
+import arcs.android.crdt.VersionMapProto
+import arcs.android.storage.ProxyMessageProto
+import java.util.concurrent.Executors
+
+/**
+ * Run asynchronously as a cache warmer to suffer from Arcs protos initialization
+ * overhead as well as cache misses on behalf of whoever needs Arcs capabilities.
+ *
+ * The initialization overhead typically comes from instantiating default instance
+ * and parsing/building proto schema.
+ *
+ * Recommend executing ArcsProtoPrefetcher() as early as possible i.e.
+ * the entry of Application lifecycle: onCreate()
+ */
+object ProtoPrefetcher {
+    /** Identify [ProtoPrefetcher] in thread lists and system traces. */
+    private const val name = "ProtoPrefetcher"
+
+    /**
+     * Single-threaded prefetcher.
+     *
+     * TODO(ianchang): tune background priority if needed.
+     */
+    private val prefecher = Executors.newFixedThreadPool(1) {
+        Thread(it).apply { name = this@ProtoPrefetcher.name }
+    }
+
+    /** Procedures to be taken to warm up Arcs protos. */
+    private val procedures = arrayOf<() -> Any>(
+        ProxyMessageProto::getDefaultInstance,
```

Comment:
Do you have any ideas for how we might keep this list up to date over time? I can imagine it will very quickly get out of date.

Reply:
I can think of 2 ways:
1) annotation processing
2) proto reflection (descriptors). Create a master proto that contains one field type of every proto in it, and then traverse the fields on this proto, using the descriptors to infer the types to reflectively build.

Reply:
I don't have a perfect way for this at this moment.
Perhaps a few new starlark rules that scan all .proto, generate temporary source files then output a **_android_library as dep.

Reply:
I am wondering if you need to explicitly invoke all the nested protos, eg have you verified if building ProxyMessageProto also warms up ProxyMessageProto.SyncRequest?
because otherwise just creating the master proto and building that would be enough

Reply:
ditto.